### PR TITLE
fix: add options arguments to esquery methods, add missing traverse method, make ancestry arg optional

### PR DESCRIPTION
**Pull request by brettz9**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/estools/esquery/blob/master/esquery.js
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/esquery/index.d.ts [hunk 1]
```diff
@@ -14,19 +15,35 @@ export = query;
...
+    }
+
     /** Parse a selector and return its AST. */
     function parse(selector: string): Selector;
     /** From a JS AST and a selector AST, collect all JS AST nodes that match the selector. */
-    function match(ast: Node, selector: Selector): Node[];
+    function match(ast: Node, selector: Selector, options?: ESQueryOptions): Node[];
     /** Given a `node` and its ancestors, determine if `node` is matched by `selector`. */
-    function matches(node: Node, selector: Selector, ancestry: Node[]): boolean;
+    function matches(node: Node, selector: Selector, ancestry?: Node[] | null, options?: ESQueryOptions): boolean;
```
Comment (cherryblossom000) [inline types/esquery/index.d.ts:30]: Why did you add the `null` here? I think just having `ancestry?: Node[]` is fine.

Although technically, the function could accept `null`:

https://github.com/estools/esquery/blob/7d1f6691d8f76a85a84214ce210427d6b055b196/esquery.js#L377

```js
    if (!ancestry) { ancestry = []; }
```

I’m not entirely sure what the guidelines about this are but it seems like to me `ancestry` is an optional parameter and I’m more inclined to type it as `ancestry?: Node[]` without the `null`.

File: types/esquery/index.d.ts [hunk 2]
```diff
@@ -14,19 +15,35 @@ export = query;
...
-    function query(ast: Node, selector: string): Node[];
+    function query(ast: Node, selector: string, options?: ESQueryOptions): Node[];
+
+    /** From a JS AST and a selector AST, collect all JS AST nodes that match the selector. */
+    function traverse(
+        ast: Node,
+        selector: string,
+        visitor: (node: Node, parent: Node, ancestry: Node[]) => void,
+        options?: ESQueryOptions,
+    ): Node[];
```
Comment (cherryblossom000) [inline types/esquery/index.d.ts:40]: The JSDoc for `traverse` say it returns a `Node[]`, but the [actual source code](https://github.com/estools/esquery/blob/7d1f6691d8f76a85a84214ce210427d6b055b196/esquery.js#L540-L579) shows that it returns `void`.

```suggestion
    ): void
```

**Review by cherryblossom000 — CHANGES_REQUESTED** [review]
Looks good, there’s just a few nitpicky things I want to point out

File: types/esquery/index.d.ts (hunk 1 above)
Comment (brettz9) [reply types/esquery/index.d.ts:30]: Besides the fact I was using it in my own code, I've been kind of a mindset that `null` should be allowed where it works and is reasonable, being both more succinct, and also given its meaning that something is consciously set to a nullish value. So I've historically hoped it should be the norm to accept `null` in such cases. However, given the extra typing it requires, perhaps it is not a convention which is likely to win out, and I'm fine dropping it.

Comment (brettz9) [reply types/esquery/index.d.ts]: I've dropped it and gotten the tests passing. Thanks!

**Review by cherryblossom000 — APPROVED** [review]

**Comment by brettz9** [comment]
Ready to merge

**Pull request merged**